### PR TITLE
[RFC] add statusline widget for back/forward indicator

### PR DESCRIPTION
**Pull request by blueyed**
Fixes https://github.com/qutebrowser/qutebrowser/issues/2737.

TODO:
- [x] tests

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/2738)

### CODE REVIEW

File: qutebrowser/mainwindow/statusbar/backforward.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+from qutebrowser.mainwindow.statusbar import textbase
+from qutebrowser.utils import log
+
+
+class Backforward(textbase.TextBase):
+
+    """Shows navigation indicator (if you can go backward and/or forward)."""
+
+    def on_tab_changed(self, tab):
+        self.tab = tab
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: I'm not entirely happy with saving the tab here, but it's a bit of a tricky situation - what about connecting the signal above like `tabs.cur_url_changed.connect(functools.partial(status.backforward.on_url_changed, tabs=tabs))`, then taking a `tabs` in `on_url_changed`, and use `tabs.currentWidget()` to get the current tab? I haven't tested it, but I think it should work and it seems less error-prone to me.

File: qutebrowser/mainwindow/statusbar/backforward.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2015-2017 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: nitpick: 2017 instead of 2015-2017

File: qutebrowser/mainwindow/statusbar/backforward.py (hunk 1 above)
Comment (blueyed) [inline qutebrowser/mainwindow/statusbar/backforward.py]: Seems to work well.

**Comment by blueyed** [comment]
I've also added the `@pyqtSlot()` decorator, not sure if it is necessary?  Maybe useful just for documentation / introspection alone?!

**Comment by blueyed** [comment]
Any pointer for the tests?
A new file in ` tests/unit/mainwindow/statusbar/`?
How can I mock / provide `tabs` there?

File: qutebrowser/mainwindow/statusbar/backforward.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+from PyQt5.QtCore import pyqtSlot
+from qutebrowser.mainwindow.statusbar import textbase
+
+
+class Backforward(textbase.TextBase):
+
+    """Shows navigation indicator (if you can go backward and/or forward)."""
+
+    @pyqtSlot()
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: `@pyqtSlot()` on methods connected to PyQt signals is usually used because it has a memory/performance benefit (and also works around some bugs in older PyQt versions).

Here you're connecting a `functools.partial` object however, and never connect to the method directly - so I'd rather remove it.

**Comment by The-Compiler** [comment]
Yeah, a new file there sounds good. When it has 100% coverage, please also add it to `scripts/dev/check_coverage.py` (like the other statusbar files there).

As for mocking `tabs`: There's a [TabbedBrowserStub](https://github.com/qutebrowser/qutebrowser/blob/master/tests/helpers/stubs.py#L558-L606) in `stubs.py`, which you can access via a `stubs` fixture.

In 994e8c692f4357a5fa0529ca8fab5c1707699d5c I just did some changes related to that as I discovered there were various other `TabbedBrowser` stubs, so I merged them into one. If you're comfortable with `git rebase`, I suggest you rebase on the newest `master` first.

**Comment by The-Compiler** [comment]
Just as a heads-up: I'm scheduled to have a (planned) eye surgery, so I probably won't respond to this anymore until somewhen next week.

**Comment by blueyed** [comment]
@The-Compiler 
Thanks for notifying me, and good luck with that!
Take care!

**Comment by blueyed** [comment]
What do you think about having some additional indicator on how far you can go forward/backwards, e.g. the number if it is `> 1`, and `< 10`, and using `<<` / `>>` in case it is `> 10`?

**Comment by The-Compiler** [comment]
Dunno. It doesn't sound too useful I think, and it's not very intuitive what it'd mean. Either way, I don't think you can get this information easily with the API we get from QtWebKit/QtWebEngine.

**Comment by The-Compiler** [comment]
Any update, @blueyed?

**Comment by blueyed** [comment]
No, I am using this myself, but otherwise I have to find the time to finish this, i.e. write the tests.

**Comment by The-Compiler** [comment]
I felt like writing some tests, so here they are:

```diff
From b6cd620a8635aad1cec325e9ef3cbfd15bec1766 Mon Sep 17 00:00:00 2001
From: Florian Bruhin <git@the-compiler.org>
Date: Wed, 5 Jul 2017 15:45:23 +0200
Subject: [PATCH 4/4] Add statusbar.backforward tests

---
 scripts/dev/check_coverage.py                      |  2 +
 tests/helpers/stubs.py                             | 20 ++++++++-
 .../unit/mainwindow/statusbar/test_backforward.py  | 50 ++++++++++++++++++++++
 3 files changed, 71 insertions(+), 1 deletion(-)
 create mode 100644 tests/unit/mainwindow/statusbar/test_backforward.py

diff --git a/scripts/dev/check_coverage.py b/scripts/dev/check_coverage.py
index 7e5a2d47a..b03fb7ef4 100644
--- a/scripts/dev/check_coverage.py
+++ b/scripts/dev/check_coverage.py
@@ -117,6 +117,8 @@ PERFECT_FILES = [
         'mainwindow/statusbar/textbase.py'),
     ('tests/unit/mainwindow/statusbar/test_url.py',
         'mainwindow/statusbar/url.py'),
+    ('tests/unit/mainwindow/statusbar/test_backforward.py',
+        'mainwindow/statusbar/backforward.py'),
     ('tests/unit/mainwindow/test_messageview.py',
         'mainwindow/messageview.py'),
 
diff --git a/tests/helpers/stubs.py b/tests/helpers/stubs.py
index 0551f5a86..788d3ace3 100644
--- a/tests/helpers/stubs.py
+++ b/tests/helpers/stubs.py
@@ -222,6 +222,22 @@ class FakeWebTabScroller(browsertab.AbstractScroller):
         return self._pos_perc
 
 
+class FakeWebTabHistory(browsertab.AbstractHistory):
+
+    def __init__(self, tab, *, can_go_back, can_go_forward):
+        super().__init__(tab)
+        self._can_go_back = can_go_back
+        self._can_go_forward = can_go_forward
+
+    def can_go_back(self):
+        assert self._can_go_back is not None
+        return self._can_go_back
+
+    def can_go_forward(self):
+        assert self._can_go_forward is not None
+        return self._can_go_forward
+
+
 class FakeWebTab(browsertab.AbstractTab):
 
     """Fake AbstractTab to use in tests."""
@@ -229,12 +245,14 @@ class FakeWebTab(browsertab.AbstractTab):
     def __init__(self, url=FakeUrl(), title='', tab_id=0, *,
                  scroll_pos_perc=(0, 0),
                  load_status=usertypes.LoadStatus.success,
-                 progress=0):
+                 progress=0, can_go_back=None, can_go_forward=None):
         super().__init__(win_id=0, mode_manager=None, private=False)
         self._load_status = load_status
         self._title = title
         self._url = url
         self._progress = progress
+        self.history = FakeWebTabHistory(self, can_go_back=can_go_back,
+                                         can_go_forward=can_go_forward)
         self.scroller = FakeWebTabScroller(self, scroll_pos_perc)
         wrapped = QWidget()
         self._layout.wrap(self, wrapped)
diff --git a/tests/unit/mainwindow/statusbar/test_backforward.py b/tests/unit/mainwindow/statusbar/test_backforward.py
new file mode 100644
index 000000000..15de7442f
--- /dev/null
+++ b/tests/unit/mainwindow/statusbar/test_backforward.py
@@ -0,0 +1,50 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2017 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+
+"""Test Progress widget."""
+
+import pytest
+
+from qutebrowser.mainwindow.statusbar import backforward
+
+
+@pytest.fixture
+def backforward_widget(qtbot):
+    widget = backforward.Backforward()
+    qtbot.add_widget(widget)
+    return widget
+
+
+@pytest.mark.parametrize('can_go_back, can_go_forward, expected_text', [
+    (False, False, ''),
+    (True, False, '<'),
+    (False, True, '>'),
+    (True, True, '<>'),
+])
+def test_backforward_widget(backforward_widget, stubs,
+                            fake_web_tab, can_go_back, can_go_forward,
+                            expected_text):
+    """Ensure the Backforward widget shows the correct text."""
+    tab = fake_web_tab(can_go_back=can_go_back, can_go_forward=can_go_forward)
+    tabbed_browser = stubs.TabbedBrowserStub()
+    tabbed_browser.current_index = 1
+    tabbed_browser.tabs = [tab]
+    backforward_widget.on_url_changed(tabbed_browser)
+    assert backforward_widget.text() == expected_text
-- 
2.13.2

```

I'll also add some more line comments.

File: qutebrowser/mainwindow/statusbar/backforward.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+from PyQt5.QtCore import pyqtSlot
+from qutebrowser.mainwindow.statusbar import textbase
+
+
+class Backforward(textbase.TextBase):
+
+    """Shows navigation indicator (if you can go backward and/or forward)."""
+
+    def on_url_changed(self, tabs):
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: It's a bit confusing still calling this `on_url_changed` I think - maybe a simple `update_text` or so would be better.

File: qutebrowser/mainwindow/statusbar/backforward.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        """Called on URL changes."""
+        tab = tabs.currentWidget()
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: On [Travis](https://travis-ci.org/qutebrowser/qutebrowser/jobs/247845777#L1399) looks like this was invoked while `tabs.currentWidget()` was `None` and caused it to raise. I'm not 100% sure why it happened, but I think it'd be a good idea to add a:

```python
if tab is None:
    return
```

File: qutebrowser/mainwindow/statusbar/backforward.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        text = ''
+        if tab.history.can_go_back():
+            text += '<'
+        if tab.history.can_go_forward():
+            text += '>'
+        self.setText(text)
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: Maybe add `[]` around the text so it fits in with the other statusbar widgets?

File: qutebrowser/mainwindow/mainwindow.py [hunk 7]
```diff
@@ -469,6 +469,7 @@ def _connect_signals(self):
         tabs.tab_index_changed.connect(status.tabindex.on_tab_index_changed)
 
         tabs.cur_url_changed.connect(status.url.set_url)
+        tabs.cur_url_changed.connect(functools.partial(status.backforward.on_url_changed, tabs=tabs))
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/mainwindow.py]: Please break this after the `partial(` or so.

File: qutebrowser/mainwindow/statusbar/backforward.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Navigation (back/forward) indicator displayed in the statusbar."""
+
+from PyQt5.QtCore import pyqtSlot
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: unused import

Comment (blueyed) [inline qutebrowser/mainwindow/statusbar/backforward.py]: Why is F401 (this error) in `.flake8`? (since the beginning in 91e6f4c37)

File: qutebrowser/mainwindow/statusbar/backforward.py (hunk 4 above)
Comment (blueyed) [inline qutebrowser/mainwindow/statusbar/backforward.py]: It is still called through `tabs.cur_url_changed`?!

File: qutebrowser/mainwindow/statusbar/backforward.py (hunk 8 above)
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: Because it's already [checked by pylint](https://travis-ci.org/qutebrowser/qutebrowser/jobs/247845780#L813) and I don't want to do ignores twice everywhere for imports which aren't actually unused.

File: qutebrowser/mainwindow/statusbar/backforward.py (hunk 6 above)
Comment (blueyed) [inline qutebrowser/mainwindow/statusbar/backforward.py]: It takes away some more space then, and IIRC the current behavior is what you see in Vimperator, too - but I will change it.

File: qutebrowser/mainwindow/statusbar/backforward.py (hunk 4 above)
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: Yes, but if something is called `on_url_changed`, I'd expect it to be directly connected to a `url_changed` signal and get a `QUrl` as argument.

File: qutebrowser/mainwindow/statusbar/backforward.py (hunk 8 above)
Comment (blueyed) [inline qutebrowser/mainwindow/statusbar/backforward.py]: Because pylint takes care of that?

Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: I don't understand what you're asking? Yes, pylint shows an unused import error, see my link above.

Comment (blueyed) [inline qutebrowser/mainwindow/statusbar/backforward.py]: Yes, pylint does, but I was using only flake8 in the editor (which I will change for the project).
I think I answered my own question: F401 from flake8 is ignored, since pylint is supposed to lint this part.

File: qutebrowser/mainwindow/statusbar/backforward.py (hunk 4 above)
Comment (blueyed) [inline qutebrowser/mainwindow/statusbar/backforward.py]: `on_tab_cur_url_changed` then maybe?
I think it is better to have some explicit name, instead of `update_text`?!

**Comment by blueyed** [comment]
@The-Compiler 
Pushed some updates.
I assume we'll squash this altogether in the end, so I've just added your tests, without your authorship?!

Remaining: the renaming of the hook.

**Comment by blueyed** [comment]
Why does Travis not run the PR?

**Comment by The-Compiler** [comment]
Good question... let me try to close and reopen it.

File: qutebrowser/mainwindow/statusbar/backforward.py (hunk 4 above)
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: Works for me :wink:

File: tests/helpers/stubs.py [hunk 9]
```diff
@@ -222,19 +222,37 @@ def pos_perc(self):
         return self._pos_perc
 
 
+class FakeWebTabHistory(browsertab.AbstractHistory):
+
```
Comment (The-Compiler) [inline tests/helpers/stubs.py]: My bad, but this needs a docstring (something like `"""Fake for Web{Kit,Engine}History."""` or so)

File: qutebrowser/mainwindow/statusbar/backforward.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class Backforward(textbase.TextBase):
+
+    """Shows navigation indicator (if you can go backward and/or forward)."""
+
+    def on_url_changed(self, tabs):
+        """Called on URL changes."""
+        tab = tabs.currentWidget()
+        if tab is None:
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/backforward.py]: Hmm, probably needs a `  # pragma: no cover` as it doesn't always happen.

**Comment by The-Compiler** [comment]
Thanks!

**Comment by blueyed** [comment]
You're welcome!
Thank you for helping me get this in!

**Comment by The-Compiler** [comment]
Pushed two small changes:

- 5fb6cb713b7047cec245ea7a3a19dddfbc1cf5aa because the gap when there's no history started to annoy me
- e81dcccacef3c241361a6bcf98e3038ec9bd53c0 to add a test for the `if tab is None:` case.
